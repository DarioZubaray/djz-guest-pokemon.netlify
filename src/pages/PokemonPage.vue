<template>
    <div id="score">
      <div>
        <span>Dificultad: </span>
        <select v-model="selected">
          <option value="2">Fácil</option>
          <option value="4">Medio</option>
          <option value="6">Dificil</option>
        </select>
      </div>
      <div>
        <small>score: </small> <span>{{score}}</span>
      </div>
    </div>

    <h1>¿Quién es este pokémon?</h1>

    <h1 v-if="!choosed">Espere por favor...</h1>
    <div v-else>
        <pokemon-image
          :pokemon-id="choosed.id"
          :show-pokemon="showPokemon"
        ></pokemon-image>
        <PokemonOptions 
          :pokemons="pokemonArr"
          :disabledButtons="showAnswer"
          @selection="checkAnswer"
        />

        <div v-if="showAnswer" class="fadeIn">
          <h2>{{ message }}</h2>

          <template v-if="!isCorrect">
            <h2 v-if="revealCorrect">{{ aditionalMessage }}</h2>
            <button @click="revealCorrectAnswer">¿Quién era?</button>
          </template>

          <button @click="newGame">
            Siguiente
          </button>
        </div>
    </div>
</template>

<script>
import PokemonImage from "@/components/PokemonImage"
import PokemonOptions from "@/components/PokemonOptions"
import getPokemonOptions from "@/helpers/getPokemonOptions"

export default {
  components: { PokemonImage, PokemonOptions },
  // props: {
  //   dificulty : {
  //       type: Number,
  //       required: false,
  //       default: 6
  //     }
  // },
  data() {
    return {
      pokemonArr: [],
      choosed: null,
      showPokemon: false,
      showAnswer: false,
      isCorrect: false,
      message: '',
      aditionalMessage: '',
      revealCorrect: false,
      score: 0,
      selected: 4,
      dificulty: 4
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions(this.dificulty)
      const randInt = Math.floor( Math.random() * this.dificulty )
      this.choosed = this.pokemonArr[ randInt ]
    },
    checkAnswer(pokemon) {
      console.log('checkAnswer', pokemon);
      this.showPokemon = true
      this.showAnswer = true
      this.isCorrect = (pokemon.id === this.choosed.id)
      if ( this.isCorrect ) this.setCorrect()
      else this.setIncorrect(pokemon.name)
    },
    setCorrect() {
      this.message = `Correcto! Es ${this.choosed.name}.`
      this.aditionalMessage = '' 
      this.score = this.score + 1 
    },
    setIncorrect(name) {
      this.message = `Oops! No era ${name}.`
      this.aditionalMessage = `Era ${this.choosed.name}.`
    },
    revealCorrectAnswer() {
      this.revealCorrect = true
    },
    newGame() {
      this.choosed =  null,
      this.showPokemon = false,
      this.showAnswer = false,
      this.isCorrect = false,
      this.message = ''
      this.aditionalMessage = ''
      this.mixPokemonArray()
    }
  },
  watch: {
    selected(value, oldValue) {
      this.dificulty = value
      this.score = 0
      this.newGame()
    }
  },
  mounted() {
    this.mixPokemonArray()
  }
}
</script>

<style>
#score {
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 10px;
}

#score span {
  font-size: 20px;
}
</style>