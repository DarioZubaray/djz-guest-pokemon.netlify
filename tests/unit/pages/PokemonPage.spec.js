import { shallowMount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { mockPokemons } from '../mocks/pokemons.mock'

describe('PokemonPage', () => {

    let wrapper

    beforeEach( () => {
        wrapper = shallowMount( PokemonPage )
    })

    test('debe de hacer match con el snapshot', () => {
      expect( wrapper.html() ).toMatchSnapshot()
    })

    test('debe de llamar el mixPokemonArray al montar', () => {
        const mixPokemonArraySPy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        wrapper = shallowMount( PokemonPage )

        expect( mixPokemonArraySPy ).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando se cargan los pokemons', () => {
        wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    choosed: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    isCorrect: false,
                    message: '',
                    aditionalMessage: '',
                    revealCorrect: false,
                    score: 0,
                    selected: 4,
                    dificulty: 4
                }
            }
        })

        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de pokemon-image y pokemon-options', () => {
      const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    choosed: mockPokemons[3],
                    showPokemon: false,
                    showAnswer: false,
                    isCorrect: false,
                    message: '',
                    aditionalMessage: '',
                    revealCorrect: false,
                    score: 0,
                    selected: 4,
                    dificulty: 4
                }
            }
        })

        
        const pokemonImage = wrapper.find('pokemon-image-stub')
        const pokemonOptions = wrapper.find('pokemon-options-stub')
        expect( pokemonImage.exists() ).toBeTruthy()
        expect( pokemonOptions.exists() ).toBeTruthy()

        expect( pokemonImage.attributes('pokemonid') ).toBe('4')
        expect( pokemonOptions.attributes('pokemons') ).toBeTruthy()
    })

    test('pruebas con checkAnswer', async() => {
        
        const wrapper = shallowMount( PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    choosed: mockPokemons[2],
                    showPokemon: false,
                    showAnswer: false,
                    isCorrect: false,
                    message: '',
                    aditionalMessage: '',
                    revealCorrect: false,
                    score: 0,
                    selected: 4,
                    dificulty: 4
                }
            }
        })

        await wrapper.vm.checkAnswer(mockPokemons[2])

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe(true)
        expect( wrapper.find('h2').text() ).toBe(`Correcto! Es ${ mockPokemons[2].name }.`)

        await wrapper.vm.checkAnswer(mockPokemons[0])

        expect( wrapper.vm.message ).toBe(`Oops! No era ${ mockPokemons[0].name }.`)

    })
})