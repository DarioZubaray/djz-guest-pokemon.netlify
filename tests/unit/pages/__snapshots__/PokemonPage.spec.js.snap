// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage debe de hacer match con el snapshot 1`] = `
<div id="score">
  <div><span>Dificultad: </span><select>
      <option value="2">Fácil</option>
      <option value="4">Medio</option>
      <option value="6">Dificil</option>
    </select></div>
  <div><small>score: </small> <span>0</span></div>
</div>
<h1>¿Quién es este pokémon?</h1>
<h1>Espere por favor...</h1>
`;

exports[`PokemonPage debe de hacer match con el snapshot cuando se cargan los pokemons 1`] = `
<div id="score">
  <div><span>Dificultad: </span><select>
      <option value="2">Fácil</option>
      <option value="4">Medio</option>
      <option value="6">Dificil</option>
    </select></div>
  <div><small>score: </small> <span>0</span></div>
</div>
<h1>¿Quién es este pokémon?</h1>
<div>
  <pokemon-image-stub pokemonid="1" showpokemon="false"></pokemon-image-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" disabledbuttons="false"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
